Source/DafnyCore/Verifier/BoogieGenerator.cs

```diff
@@ -3884,58 +3885,81 @@ public ForceCheckToken(IToken tok)
       return isAlloc == null ? isPred : isPred == null ? isAlloc : BplAnd(isPred, isAlloc);
     }
 
-    // Returns an expression, which, if false, means that the two LHS expressions are
+    // Returns expressions, which, if false, means that the two LHS expressions are
     // not distinct; if null then the LHSs are trivially distinct
-    Bpl.Expr CheckDistinctness(Expression lhsa, Expression lhsb, ExpressionTranslator etran) {
+    void CheckDistinctness(Expression lhsa, Expression lhsb, ExpressionTranslator etran, out Expression dExpr, out Bpl.Expr bExpr) {
+      // Helper functions to make the following more concise
+      Expression DfyBin(BinaryExpr.Opcode op, Expression e0, Expression e1) =>
```

Comment:
I think the existing static method `Expression.CreateOr` and its ilk would make these unnecessary.

Reply:
Summarizing our Slack discussion - I've been leaning towards these lower-level constructions since they work regardless of the constituent expressions' resolved state (which can be hard to predict, now and in the future). It would be nice to have resolved expressions everywhere and to use the helper methods, but I'll leave that for another time.